### PR TITLE
feat(module:date-picker): support parsing input value

### LINKED ISSUE #2492: Problem to set date manually datepicker when chage nzFormat
## Version
1.8.0
## Environment
Windows
## Reproduction link
[https://ng-zorro-antd-start-cekrxg.stackblitz.io](https://ng-zorro-antd-start-cekrxg.stackblitz.io)
## Steps to reproduce
Set dateformat to yyyy-MM-dd in the nz-datepicket, after set date manually in the input box datepicker a date for example 2018-01-01
## What is expected?
Ever
## What is actually happening?
Problem to set manually date when dateFormat is changed. For example, if you set nzFormat to yyyy/MM/dd set date manually work fine, the datepicker set a date like 2018/01/01 then hide, this is ok. But when nzFormat is set to yyyy-MM-dd, set date manually not work, the datepicker is maintain visible and if you does click outside the calendar the date is not set

### LINKED ISSUE #3976: DatePicker doesn't recognize manual date input when using nzFormat
### Reproduction link
[https://stackblitz.com/edit/angular-a8d35n](https://stackblitz.com/edit/angular-a8d35n)

### Steps to reproduce
Add DatePicker component and set nzFormat="dd.MM.yyyy".
```javascript
<nz-date-picker [(ngModel)]="date" nzFormat="dd.MM.yyyy"></nz-date-picker>
```
Manual input works only if date and month are same. For example, if we type 04.04.1995 it works, but 04.05.1995 is not working.

### What is expected?
After writing string date, DatePicker changes date in the model.

### What is actually happening?
Nothing happens

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.1.2 |
| Browser | Chrome 76 |

---
Default format working fine

### LINKED ISSUE #4028: Enter key selects 2001 on nz-date-picker
### Reproduction link
[https://stackblitz.com/edit/ng-zorro-antd-start-ljdvgm](https://stackblitz.com/edit/ng-zorro-antd-start-ljdvgm)

### Steps to reproduce
1. Click date-picker input box
2. Pick date
3. Click 'ok'
4. Click date-picker input box (again)
5. Press 'enter' key. Notice year changes to 2001.

### What is expected?
Enter key will close date-picker.

### What is actually happening?
Enter changes year 2001.

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.2.0 |
| Browser | Version 76.0.3809.100 (Official Build) (64-bit) |

### LINKED ISSUE #4101: 日期选择框最上方的input框修改日期无效
### Reproduction link
[https://ng-zorro-antd-start.stackblitz.io](https://ng-zorro-antd-start.stackblitz.io)

### Steps to reproduce
点击日期选择框，然后手动修改上方input里面的日期，点击确认时间不变

### What is expected?
时间变成最新修改的时间

### What is actually happening?
时间还是自己选中的时间，不是修改以后的时间

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.2.1 |
| Browser |  76.0.3809.132 |

### PR DESCRIPTION
**Pull request by wenqi73**
Close #4028 
Close #3976
Close #2492
Close #4101

~~默认的日期格式使用国际化文件中的 [`dateFormat`](https://github.com/NG-ZORRO/ng-zorro-antd/blob/cc3868ee340e7d8abd34a12664ac3c3b34510cf3/components/i18n/languages/calendar/en_US.ts#L24) 属性对应的值。~~

## 新功能
date-fns 1.x 不支持日期的反格式化，相关 issue 有：#4028 #3976 #2492 #4101。
date-fnx 2.x 已支持[该功能](https://date-fns.org/v2.9.0/docs/parse)。

## 如何实现
由于 angular 的[日期格式](https://angular.io/api/common/DatePipe)与 date-fns 的[日期格式](https://date-fns.org/v2.9.0/docs/parse)并不兼容。因此反格式化功能只有在用户引入了 date-fns 才能使用。

## Breaking Change
Changelog: https://github.com/date-fns/date-fns/blob/master/CHANGELOG.md#200---2019-08-20
我们建议用户同步 NG-ZORRO 升级到 2.x 版本。

## PR Checklist
Please check if your PR fulfills the requirements:

- [ ] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information

API has changed, please merge in 9.0.0.

### CODE REVIEW

**Comment by daxilsshah** [comment]
@wenqi73 Any updates on this PR? we're waiting for the release of these changes.

**Comment by wenqi73** [comment]
@daxilsshah We are evaluating the breaking changes this feature and will give some feedback here.

File: components/date-picker/abstract-picker.component.ts [hunk 1]
```diff
@@ -176,9 +176,12 @@ export abstract class AbstractPickerComponent implements OnInit, OnChanges, OnDe
   private setLocale(): void {
     this.nzLocale = this.i18n.getLocaleData('DatePicker', {});
     this.setDefaultPlaceHolder();
+    this.setFormat();
     this.cdr.markForCheck();
   }
 
+  protected setFormat(): void {}
```
Comment (wendzhue) [inline components/date-picker/abstract-picker.component.ts]: `abstract`

File: components/date-picker/demo/format.ts [hunk 2]
```diff
@@ -21,6 +21,6 @@ import { Component } from '@angular/core';
   ]
 })
 export class NzDemoDatePickerFormatComponent {
-  dateFormat = 'yyyy/MM/dd';
-  monthFormat = 'yyyy/MM';
+  dateFormat = 'dd.MM.yyyy';
```
Comment (wendzhue) [inline components/date-picker/demo/format.ts]: demo 可以不变吗？

**Comment by zaideygrek** [comment]
Hello, what about this fix, do you need help?

File: components/date-picker/abstract-picker.component.ts (hunk 1 above)
Comment (wenqi73) [reply components/date-picker/abstract-picker.component.ts]: 用了 `implements` 后就不能用 `abstract` 了。

**Comment by dylanvdmerwe** [comment]
Very interested in this PR.

**Comment by wenqi73** [comment]
See #4833

**Pull request closed without merging**